Copy index file content to c memory

congqixia opened the pull request:
Might be the root cause of #15896
The index content bytes were passed via unsafe.Pointer,
which shall be copy to cpp managed memory inmediately

This solution might cause a lot of memory consumption.
Maybe we need to store the index bytes in cpp memory in the first place.

/kind bug
/cc @czs007 

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

congqixia commented:
/hold

DragonDriver commented on internal/core/src/common/vector_index_c.cpp:
```diff
@@ -47,7 +47,9 @@ AppendIndexBinary(CBinarySet c_binary_set, void* index_binary, int64_t index_siz
-        std::shared_ptr<uint8_t[]> data(index, [](void*) {});
+        uint8_t* dup = new uint8_t[index_size]();
+        memcpy(dup, index, index_size);
+        std::shared_ptr<uint8_t[]> data(dup, [](void*) {});
```
`dup` here won't be released.

congqixia committed:
Copy index file content to c memory

Might be the root cause of #15896
The index content bytes were passed via unsafe.Pointer,
which shall be copy to cpp managed memory inmediately

This solution might cause a lot of memory consumption.
Maybe we need to store the index bytes in cpp memory in the first place.

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

cydrain commented:
/lgtm

xiaofan-luan commented:
/approve

congqixia commented:
/unhold
